[WIP] physics informed neural operator ode (#806)
SciML/NeuralPDE.jl: docs/src/tutorials/pino_ode.md

* output data: set of solutions u(t){a} corresponding parameter 'a'.

```@example pino
t0, t_end = tspan
instances_size = 50
range_ = range(t0, stop = t_end, length = instances_size)
ts = reshape(collect(range_), 1, instances_size)
batch_size = 50
as = [Float32(i) for i in range(0.1, stop = pi / 2, length = batch_size)]

u_output_ = zeros(Float32, 1, instances_size, batch_size)
prob_set = []
for (i, a_i) in enumerate(as)
    prob_ = ODEProblem(ODEFunction(linear, analytic = linear_analytic), u0, tspan, a_i)
    sol1 = solve(prob_, Tsit5(); saveat = 0.0204)
    reshape_sol = Float32.(reshape(sol1(range_).u', 1, instances_size, 1))
    push!(prob_set, prob_)
    u_output_[:, :, i] = reshape_sol
end
train_set = TRAINSET(prob_set, u_output_)
```

Here it used the PINO method to learning operator of the given family of parametric ODEs. 

```@example pino
chain = Lux.Chain(Lux.Dense(2, 16, Lux.σ),
    Lux.Dense(16, 16, Lux.σ),
    Lux.Dense(16, 16, Lux.σ),
    Lux.Dense(16, 32, Lux.σ),
    Lux.Dense(32, 32, Lux.σ),
    Lux.Dense(32, 1))
# flat_no = FourierNeuralOperator(ch = (2, 16, 16, 16, 16, 16, 32, 1), modes = (16,),
#     σ = gelu)

opt = OptimizationOptimisers.Adam(0.01)

> pino_phase = OperatorLearning(train_set, is_data_loss = true, is_physics_loss = true)

Reviewer: Why is a training dataset required here? That's just the normal neural operator. Can you show the pure PINO?

Author: usually what I've seen PINO mean train with both and data and physics loss.

Reviewer: But it shouldn't be required.

alg = PINOODE(chain, opt, pino_phase)
pino_solution = solve(
    prob, alg, verbose = false, maxiters = 3000)
predict = pino_solution.predict
ground = u_output_
```

Now let's compare the predictions from the learned operator with the ground truth solution which is obtained early by numerically solving the parametric ODE. Where 'i' is the index of the parameter 'a' in the dataset.

```@example pino
using Plots
i=45
plot(predict[1, :, i], label = "Predicted")
plot!(ground[1, :, i], label = "Ground truth")
```

Now to move on the stage of solving a certain equation using a trained operator and physics

## Solve ODE using learned operator family of parametric ODE for fine tuning.
```@example pino
dt = (t_end - t0) / instances_size
pino_phase = EquationSolving(dt, pino_solution)
chain = Lux.Chain(Lux.Dense(2, 16, Lux.σ),
    Lux.Dense(16, 16, Lux.σ),
    Lux.Dense(16, 32, Lux.σ),
    Lux.Dense(32, 1))
alg = PINOODE(chain, opt, pino_phase)

> fine_tune_solution = solve( prob, alg, verbose = false, maxiters = 2000)

Reviewer: I don't understand why this one is needed. Why not just use the `pino_solution `?

Author: 'pino_solution' is prediction for all family of parametric ODE. 'fine_tune_solution' is prediction for only one instance of ODE from all family. So we do additional training but for predict only one exemplar and increase accuracy using  already trained operator from 'pino_solution'.
'pino_solution' have trained for all but less accurate,  'fine_tune_solution'  for one but with better accurate then 'pino_solution'

using Lux
using Statistics, Random
using NeuralOperators
using NeuralPDE

linear_analytic = (u0, p, t) -> u0 + sin(p * t) / (p)
linear = (u, p, t) -> cos(p * t)
tspan = (0.0f0, 2.0f0)
u0 = 0.0f0
```

Generate a dataset for learning a given family of ODEs where the parameter 'a' is varied. The dataset is generated by solving the ODE for different values of 'a' and storing the solutions. The dataset is then used to train the PINO model:
* input data: set of parameters 'a',
* output data: set of solutions u(t){a} corresponding parameter 'a'.

```@example pino
t0, t_end = tspan
instances_size = 50
range_ = range(t0, stop = t_end, length = instances_size)
ts = reshape(collect(range_), 1, instances_size)
batch_size = 50
as = [Float32(i) for i in range(0.1, stop = pi / 2, length = batch_size)]

u_output_ = zeros(Float32, 1, instances_size, batch_size)
prob_set = []
for (i, a_i) in enumerate(as)
    prob = ODEProblem(ODEFunction(linear, analytic = linear_analytic), u0, tspan, a_i)
    sol1 = solve(prob, Tsit5(); saveat = 0.0204)
    reshape_sol = Float32.(reshape(sol1(range_).u', 1, instances_size, 1))
    push!(prob_set, prob)
    u_output_[:, :, i] = reshape_sol
end

> train_set = TRAINSET(prob_set, u_output_);

Author: Yes, it is quite possible to learn nn for a family of parametric equations use only physics loss. In the same way, as in a regular pinns. There is only one more dimension as a parameters.

> train_set = TRAINSET(prob_set, u_output_);

Author: Although Fourier neural operator it look like it's not really suitable for this role(learn only on physics operator) and It was first of all design for learning from data.

```

Here it used the PINO method to train the given family of parametric ODEs. 

```@example pino
prob = ODEProblem(linear, u0, tspan, 0)
flat_no = FourierNeuralOperator(ch = (2, 16, 16, 16, 16, 16, 32, 1), modes = (16,),
    σ = gelu)
opt = OptimizationOptimisers.Adam(0.03)
alg = PINOODE(flat_no, opt, train_set; is_data_loss = true, is_physics_loss = true)
pino_solution = solve(prob, alg, verbose = false, maxiters = 1000)
predict = pino_solution.predict

> ground = u_output_

Reviewer: Shouldn't it give you a function of the parameters or initial conditions or controls to then re-evaluate?

Author: I forgot about fine-tuning instances. I'll add the feature.

# Physics Informed Neural Operator for ODEs Solvers

This tutorial provides an example of how to use the Physics Informed Neural Operator (PINO) for solving a family of parametric ordinary differential equations (ODEs).

## Operator Learning  for a family of parametric ODE.

In this section, we will define a parametric ODE and solve it using a PINO. The PINO will be trained to learn the mapping from the parameters of the ODE to its solution.

```@example pino
using Test
using OrdinaryDiffEq, OptimizationOptimisers
using Lux
using Statistics, Random
using NeuralPDE

equation = (u, p, t) -> cos(p * t)
tspan = (0.0f0, 1.0f0)
u0 = 1.0f0
prob = ODEProblem(equation, u0, tspan)

# Define the architecture of the neural network that will be used as the PINO.
branch = Lux.Chain(
    Lux.Dense(1, 10, Lux.tanh_fast),
    Lux.Dense(10, 10, Lux.tanh_fast),
    Lux.Dense(10, 10))
trunk = Lux.Chain(
    Lux.Dense(1, 10, Lux.tanh_fast),
    Lux.Dense(10, 10, Lux.tanh_fast),
    Lux.Dense(10, 10, Lux.tanh_fast))
deeponet = NeuralPDE.DeepONet(branch, trunk; linear = nothing)

> bounds = (p = [0.1f0, pi],)

Reviewer: Having more than one parameters would be more illustrative of it's use case. Right now branch and trunk both have size one input and that makes it potentially confusing to the user for how to modify this demo towards a case with more parameters

Author: Ok, I will do. It will take some time to implement this feature for many parameters.

db = (bounds.p[2] - bounds.p[1]) / 50
dt = (tspan[2] - tspan[1]) / 40

> strategy = NeuralPDE.GridTraining([db, dt])

Reviewer: Don't use grid training

Author: Ok, I will support also QuasiRandomTraining for PINO ODE and use it in Doc  example

Author: StochasticTraining*

opt = OptimizationOptimisers.Adam(0.03)
alg = NeuralPDE.PINOODE(deeponet, opt, bounds; strategy = strategy)
sol = solve(prob, alg, verbose = false, maxiters = 2000)

> predict = sol.u

Reviewer: sol.original is needed to explain

```

Now let's compare the prediction from the learned operator with the ground truth solution which is obtained by analytic solution the parametric ODE. Where 
Compare prediction with ground truth.

```@example pino
using Plots
# Compute the ground truth solution for each parameter
ground_analytic = (u0, p, t) -> u0 + sin(p * t) / (p)

> p_ = bounds.p[1]:strategy.dx[1]:bounds.p[2]

Reviewer: Not using grid training will make this more compelling since it should predict at new parameters, not ones trained on

Author: yes,I agree

p = reshape(p_, 1, size(p_)[1], 1)
ground_solution = ground_analytic.(u0, p, sol.t.trunk)

# Plot the predicted solution and the ground truth solution as a filled contour plot
# sol.u[1, :, :], represents the predicted solution for each parameter value and time
plot(predict[1, :, :], linetype = :contourf)

> plot!(ground_solution[1, :, :], linetype = :contourf)

Reviewer: You don't show how to generate the solution at new parameters which is the key to the pini interface

Author: as new parameters you means some another mesh that don't use for training but in the same boundary of parameters?

Reviewer: yes

# Physics Informed Neural Operator for ODEs

This tutorial provides an example of how to use the Physics Informed Neural Operator (PINO) for solving a family of parametric ordinary differential equations (ODEs).

## Operator Learning for a family of parametric ODEs

> In this section, we will define a parametric ODE and solve it using a PINO. The PINO will be trained to learn the mapping from the parameters of the ODE to its solution.

Reviewer 2: ```suggestion
In this section, we will define a parametric ODE and then learn it with a PINO using [`PINOODE`](@ref). The PINO will be trained to learn the mapping from the parameters of the ODE to its solution.
```

```@example pino

> using Test

Reviewer 2: I don't think it is resolved.

using OptimizationOptimisers
using Lux
using Statistics, Random
using LuxNeuralOperators
using NeuralPDE

equation = (u, p, t) -> p[1] * cos(p[2] * t) + p[3]

> tspan = (0.0f0, 1.0f0)

Author: why?

> tspan = (0.0f0, 1.0f0)

Reviewer 2: All PINNs/DeepONet etc based are best trained with high precision. IMO, our defaults should be Float64.

u0 = 1.0f0
prob = ODEProblem(equation, u0, tspan)

number_of_parameter = 3
deeponet = LuxNeuralOperators.DeepONet(
    Chain(
        Dense(number_of_parameter => 10, Lux.tanh_fast), Dense(10 => 10, Lux.tanh_fast), Dense(10 => 10)),
    Chain(Dense(1 => 10, Lux.tanh_fast), Dense(10 => 10, Lux.tanh_fast),
        Dense(10 => 10, Lux.tanh_fast)))

u = rand(3, 50)
v = rand(1, 40, 1)
θ, st = Lux.setup(Random.default_rng(), deeponet)
c = deeponet((u, v), θ, st)[1]

bounds = [(1.0f0, pi), (1.0f0, 2.0f0), (2.0f0, 3.0f0)]

> number_of_parameters = 50

Reviewer 2: ```suggestion
number_of_parameter_samples = 50
```

to avoid confusion from `number_of_parameter`

strategy = StochasticTraining(40)
opt = OptimizationOptimisers.Adam(0.03)

> alg = PINOODE(deeponet, opt, bounds, number_of_parameters; strategy = strategy)

Reviewer 2: I think this needs to explained step by step as its not intuitive to a user. So breakdown step by step with some small explanations.

> sol = solve(prob, alg, verbose = true, maxiters = 3000)

Reviewer 2: ```suggestion
sol = solve(prob, alg, verbose = false, maxiters = 3000)
```

better for users and CI

Reviewer 2: This too is not resolved

```

> Now let's compare the prediction from the learned operator with the ground truth solution which is obtained by analytic solution the parametric ODE. Where 
> Compare prediction with ground truth.

Reviewer 2: ```suggestion
Now let's compare the prediction from the learned operator with the ground truth solution which is obtained by analytic solution of the parametric ODE.
```

```@example pino

> using Plots

Reviewer 2: Same here, add couple of lines about what is being visualized.

function get_trainset(bounds, tspan , number_of_parameters, dt)
    p_ = [range(start = b[1], length = number_of_parameters, stop = b[2]) for b in bounds]
    p = vcat([collect(reshape(p_i, 1, size(p_i,1))) for p_i in p_]...)
    t_ = collect(tspan[1]:dt:tspan[2])
    t = collect(reshape(t_, 1, size(t_, 1), 1))
    (p,t)
end

# Compute the ground truth solution for each parameter
ground_solution = (u0, p, t) -> u0 + p[1] / p[2] * sin(p[2] * t) + p[3]*t
function ground_solution_f(p,t)
    reduce(hcat,[[ground_solution(u0, p[:, i], t[j]) for j in axes(t, 2)] for i in axes(p, 2)])
end

(p,t) = get_trainset(bounds, tspan, 50, 0.025f0)
ground_solution_ = ground_solution_f(p,t)
predict = sol.interp((p,t))

> # Calculate the mean error and  the standard deviation of the errors

Reviewer 2: ```suggestion
# Calculate the mean error and the standard deviation of the errors
```

errors = ground_solution_ - predict
mean_error = mean(errors)
std_error = std(errors)

# generate the solution with new parameters for test the model
p,t = get_trainset(bounds, tspan, 100,  0.01f0) 
ground_solution_ = ground_solution_f(p,t)
predict = sol.interp((p,t))

errors = ground_solution_ - predict
mean_error = mean(errors)
std_error = std(errors)

# Plot the predicted solution and the ground truth solution as a filled contour plot
# predict, represents the predicted solution for each parameter value and time
plot(predict, linetype = :contourf)
plot!(ground_solution_, linetype = :contourf)
```

```@example pino
# 'i' is the index of the parameter 'p' in the dataset 
i = 5
# 'predict' is the predicted solution from the PINO model
plot(predict[:, i], label = "Predicted")
# 'ground' is the ground truth solution
plot!(ground_solution_[:, i], label = "Ground truth")
```

> 

Reviewer 2: ```suggestion
```
